wallet-new/server/Cardano/Wallet/API/V1/Handlers/Wallets.hs

```diff
@@ -2,30 +2,67 @@ module Cardano.Wallet.API.V1.Handlers.Wallets where
 
 import           Universum
 
+import qualified Pos.Wallet.Web.ClientTypes.Types        as V0
+import qualified Pos.Wallet.Web.Methods.Restore          as V0
+
 import qualified Cardano.Wallet.API.V1.Handlers.Accounts as Accounts
-import           Cardano.Wallet.API.V1.Types
+import           Cardano.Wallet.API.V1.Migration
+import           Cardano.Wallet.API.V1.Types             as V1
 import qualified Cardano.Wallet.API.V1.Wallets           as Wallets
 
+import qualified Data.ByteArray                          as ByteArray
+import qualified Data.ByteString                         as BS
+import           Data.Default                            (def)
+import           Formatting                              (int, sformat, shown, (%))
+import           Pos.Crypto.Signing.Types                (PassPhrase, passphraseLength)
+import           Pos.Wallet.Web.Methods.Logic            (MonadWalletLogic)
+import qualified Serokell.Util.Base16                    as Base16
 import           Servant
 import           Test.QuickCheck                         (arbitrary, generate, resize)
 
-handlers :: Server Wallets.API
+-- | All the @Servant@ handlers for wallet-specific operations.
+handlers :: ( HasConfiguration
+            , HasCompileInfo
+            , HasInfraConfiguration
+            , HasSscConfiguration
+            ) => ServerT Wallets.API MonadV1
 handlers =   newWallet
         :<|> listWallets
         :<|> (\walletId -> do
                      updatePassword walletId
                 :<|> deleteWallet walletId
                 :<|> getWallet walletId
                 :<|> updateWallet walletId
-                :<|> Accounts.handlers walletId
+                -- :<|> Accounts.handlers walletId
              )
 
-newWallet :: NewWallet -> Handler Wallet
-newWallet _ = liftIO $ generate arbitrary
+-- | TODO(adinapoli): Move this into a more specific module eventually.
+-- For now we are just interested in backporting features without proper
+-- error handling.
+mkPassPhrase :: SpendingPassword -> Either Text PassPhrase
+mkPassPhrase text = do
+  bs <- Base16.decode text
+  let bl = BS.length bs
+  -- Currently passphrase may be either 32-byte long or empty (for
+  -- unencrypted keys).
+  if bl == 0 || bl == passphraseLength
+      then pure $ ByteArray.convert bs
+      else fail . toString $ sformat
+           ("Expected password length 0 or "%int%", not "%int)
+           passphraseLength bl
```

Comment:
I'm strongly for getting rid of proxy types like `SpendingPassword` completely and instead providing proper JSON instances for target datatypes like `PassPhrase`, and including validation logic in these JSON instances. Proxy types (like `CWallet`, etc. in old wallet) are pain and they were initially introduced to make autogenerated code for purescript-bridge possible, but we're not going to use purescript-bridge with API V1 anymore.

Reply:
> providing proper JSON instances for target datatypes like PassPhrase, and including validation logic in these JSON instances

I see your point but I tend to disagree just because I think it's not their place. JSON instances should do just one thing well, which is making sure the input data is valid JSON, nothing else. Overloading them with too many responsibility feels like the wrong design choice. What do you think? 😉 

I do see why you find this appealing, which is because this way your handlers look nice and clean, but I would much rather create some extra layer in our handlers to perform the data validation than overloading very specialised instances with too many responsibilities.

Reply:
I just had a quick chat with @adinapoli-iohk . I explained that my view is that one should probably decide this on a case by case basis. If the external syntax & validation is fixed and the validation is cheap then having just the one type is fine. If the validation is expensive and one needs to be able to read without validating, or if there are multiple different external syntaxes, or the validation is complex or needs additional input, then a two-step approach is better. For example with decoding blocks from the network we want a two-step process, to avoid the network layer having to do the expensive validation, and have that be done only by the logic layer.

Reply:
@dcoutts I completely agree, and I think that such types as `Coin`, `PassPhrase` and `Address` go to the first category, Reasons:
- Validation for them is simple and cheap (for `Address` there's actually no validation in JSON instance directly, because JSON representation of address is its binary representation encoded in base58, so we simply reuse validation logic from `Bi` instance)
- We never need to access textual representation of these types directly in code (only to log them, but we have `Buildable` instances for that)

I cannot actually come up right away with any type which is transferred via HTTP API and which goes into second category. 

Regarding your words, @adinapoli-iohk:
> JSON instances should do just one thing well, which is making sure the input data is valid JSON, nothing else.

No, JSON instances should make sure that input data _is a valid JSON representation of value of given type_. Consider the following example: say you have this datatype:
```
data Foo = { bar :: Int, baz :: String }
```

Now your endpoint accepts `Foo` as a request body. Then you have a query with a following request body:
```
{}
```
Or this one:
```
{"patak": {"bardaq": "skovoroda"}}
```
Are these inputs valid JSON? Definitely. Does it mean that JSON instance for `Foo` which tries to construct `Foo` object from these JSON inputs shouldn't throw an error? No, because a valid `Foo` object cannot be meaningfully constructed given either of these JSON inputs.

It's the same with such types as `PassPhrase` or `Address`: JSON instance need to check whether or not given JSON represents valid value of these datatypes -- nothing else.

Reply:
@flyingleafe First of all, thanks for the nice reply. I think a healthy debate is always the best way to reach the best design possible. Let me start from the bottom! 

> JSON instances should make sure that input data is a valid JSON representation of value of given type.

Yes, and I'm afraid I have spectacularly failed in my attempt to explain myself. What you say is correct, but what I meant was that one may argue (in some cases) that your input type should respect the "type footprint" without doing any validation inside. Let's take the example you posted: we all agree that those two raw JSONs shouldn't be reified into `Foo`, but if you give me `{"bar": 42, "baz": "foo" }` now it's a matter of whether or not you want to embed special validation logic to reject this input because, say, `Foo` should have `bar` always less than 50. So I hope now is a bit clearer why a was a bit on the fence in allowing validation in our Aeson instances, and that's because there is always the risk of restricting the domain of our types too much.

After speaking with @dcoutts though, I also see the other side of the coin, which is that these Aeson instances are really parsers, and parsers usually come with semantic attached to them (for example, if I was rolling the `FromJSON PassPhrase` parser by hand, I would have surely had some parsing rule to check if the passphrase length was 32 bits long, so that's essentially the same thing that you are proposing here).

> I cannot actually come up right away with any type which is transferred via HTTP API and which goes into second category.

I can think of one! 😉  That would be a `BackupPhrase`. Whilst we might want to embed validation logic into our `FromJSON` instance to check that the words are BIP-39 compliant, surely we don't want to embed validation logic on its _length_, because that might change in the future due to requirements; today is 12 words long, for an external wallet can be 24, for other cases can be 15. By embedding _too much_ validation logic we are constraining the subset of valid types too much, imho. Nothing which cannot be overcome by newtypes, mind you, but I still think this is an example of where we might want to check some invariant but not necessarily all of them.

The TL;DR is that I'm on board with the validation logic into `Aeson` instances, provided we do this with caution and we assess how much to validate on a case-by-case  basis. Do you think you can leave with that? 😉

Reply:
@adinapoli-iohk I agree with the example of `BackupPhrase`, and I'm glad that we've seemingly reached the consensus here :)

Reply:
Of course we are! :) Ok, leave with me, I will amen this PR accordingly with all we have said here so far.